lib/datadog/tracing/contrib/pg/instrumentation.rb

```diff
@@ -101,9 +104,18 @@ def trace(name, sql: nil, statement_name: nil)
                   )
                 end
 
-                result = yield(propagated_sql_statement)
-                annotate_span_with_result!(span, result)
-                result
+                # Read metadata from PG::Result
+                if block
+                  yield(propagated_sql_statement, proc do |result|
+                    ret = block.call(result)
+                    annotate_span_with_result!(span, result)
```

Comment:
> If the optional code block is given, it will be passed result as an argument, and the [PG::Result](https://deveiate.org/code/pg/PG/Result.html) object will automatically be cleared when the block terminates. In this instance, conn.exec returns the value of the block.

According to the documentation, the `PG::Result` would be garbage collected once the block ended. I think this implementation is very nice to keep the same promise.

Since the `result` is already available for annotation, would it make sense to annotate before invoking the given block?

```
annotate_span_with_result!(span, result)
block.call(result)
```